### PR TITLE
Add PL_shutdownhook

### PR DESCRIPTION
**Pull request by Leont**
This is a proposed solution to #19022

This adds a hook that is run early in `PERL_SYS_TERM` to terminate global resources, and uses it to destroy the shared interpreter in `threads::shared`

### CODE REVIEW

**Comment by iabyn** [comment]
On Fri, Aug 06, 2021 at 07:57:06AM -0700, Leon Timmermans wrote:
> This is a proposed solution to #19022
> 
> This adds a hook that is run early in `PERL_SYS_TERM` to terminate global resources, and uses it to destroy the shared interpreter in `threads::shared`
> You can view, comment on, or merge this pull request online at:

Any particular reason PL_shutdownhook  is manually declared in perl.h
rather than being in perlvars.h like other global vars?

-- 
Never do today what you can put off till tomorrow.

**Comment by Leont** [comment]
> Any particular reason PL_shutdownhook is manually declared in perl.h rather than being in perlvars.h like other global vars?

I didn't know about that, but that does seem like a better idea. I copied it from something else that also wasn't in perlvars.h (but I forgot what)

**Comment by Leont** [comment]
> Any particular reason PL_shutdownhook is manually declared in perl.h rather than being in perlvars.h like other global vars?

Fixed that.

File: dosish.h
Comment (leonerd) [inline dosish.h]: Most minor of nitpicks, but can we reflow these words and reälign the resulting `\` escapes, to keep it neat?

**Comment by leonerd** [comment]
Aside from my comment about reflowing a macro, this LGTM

File: dosish.h
Comment (Leont) [reply dosish.h]: We could, but to be honest I was about to open a topic about restructuring `PERL_SYS_TERM_BODY` entirely. There are several definitions of it, but they're 90% identical. That seems suboptimal.

**Comment by tonycoz** [comment]
could you please document PL_shutdownhook?

**Comment by khwilliamson** [comment]
So, are we just waiting for @Leont  to document this before it is ready to merge?

**Comment by tonycoz** [comment]
```
* leont is not sure where/how to document PL_shutdownhook, given that 
          none of the other similar hooks are documented.
```
I'm not so worried about where, more about adding another undocumented public hook.

Things to document:
- when is the hook called?  Do I need to chain to the old hook?
- what do I need to lock to set it? (though a simple API to set (or add/remove a given function pointer) it might be better, since hooks have led to some bugs in the past)

**Comment by khwilliamson** [comment]
@leont, this looks close to being ready

**Comment by jkeenan** [comment]
> @Leont, this looks close to being ready

@LeonT, can we get an update on the status of this pull request?

**Comment by Leont** [comment]
> @Leont, can we get an update on the status of this pull request?

I'm adapting it to tony's suggestion (simple API to set it), will update it soon

**Comment by demerphq** [comment]
@leont last message in this PR is that you were adapting it to feedback. Do you still want to keep this PR? (It seems reasonable to me in principal).

**Comment by Leont** [comment]
> Do you still want to keep this PR? (It seems reasonable to me in principal).

Yeah, it's a problem that needs a solution. I'll try to pick it up again soon.

**Comment by demerphq** [comment]
@Leont I'm happy to get it rebased for you considering you seem busy of late.

**Comment by Leont** [comment]
> I'm happy to get it rebased for you considering you seem busy of late.

It needs much more than a rebase, otherwise this would have been merged already. I'll try to finish this soon.

**Comment by jkeenan** [comment]
> > I'm happy to get it rebased for you considering you seem busy of late.
> 
> It needs much more than a rebase, otherwise this would have been merged already. I'll try to finish this soon.

@Leont, have you been able to move any of this work forward? Thanks.

**Comment by jkeenan** [comment]
Needs a version bump; otherwise, progressing well.

**Comment by Leont** [comment]
> I'm not so worried about where, more about adding another undocumented public hook.

I've been thinking about how to make this prettier, but I now think we should make it non-public instead.

This is needed for `threads::shared`, and `threads::shared` only. Just let it be a private thing.

**Comment by tonycoz** [comment]
> > I'm not so worried about where, more about adding another undocumented public hook.
> 
> I've been thinking about how to make this prettier, but I now think we should make it non-public instead.
> 
> This is needed for `threads::shared`, and `threads::shared` only. Just let it be a private thing.

That's fine, as long as we document that: don't use this, if you do and it breaks you get to keep the pieces.

**Review by tonycoz — APPROVED** [review]

**Pull request merged**
